### PR TITLE
ENH: Allow pairwise calcuation when comparing the column with itself …

### LINKED ISSUE #25781: Using corr with callable gives 1 on diagonals where the result should be NaN
#### Code Sample, a copy-pastable example if possible

```python
import numpy as np
import pandas as pd
from scipy.stats import pearsonr

df = pd.DataFrame(np.zeros((2,2)))

df.corr(method=lambda x, y: pearsonr(x, y)[0])
```
gives:
```python
     0    1
0  1.0  NaN
1  NaN  1.0
```

#### Problem description

This is a follow up on #25726. In the example above the self-correlation is returned as 1.0. However, it should actually be `nan`. I would argue the diagonal elements of the correlation matrix should be calculated instead of setting them to 1 here:
https://github.com/pandas-dev/pandas/blob/cb00deb94500205fcb27a33cc1d0df79a9727f8b/pandas/core/frame.py#L7025-L7026

`df.corr()` gives the output I expected:

#### Expected Output

```python
df.corr()
```
```python
    0   1
0 NaN NaN
1 NaN NaN
```
#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.8.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.165-81-default
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: en_US.UTF-8
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.24.2
pytest: 4.1.1
pip: 18.1
setuptools: 40.6.3
Cython: 0.29.3
numpy: 1.15.4
scipy: 1.2.0
pyarrow: None
xarray: None
IPython: 7.2.0
sphinx: 1.8.3
patsy: 0.5.1
dateutil: 2.7.5
pytz: 2018.9
blosc: None
bottleneck: None
tables: 3.4.4
numexpr: 2.6.9
feather: None
matplotlib: 3.0.2
openpyxl: 2.4.0-b1
xlrd: 1.2.0
xlwt: None
xlsxwriter: None
lxml.etree: 4.3.0
bs4: 4.7.1
html5lib: 1.0.1
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None

</details>

### PR DESCRIPTION
**Pull request by peterpanmj**
…(#25781)

- [x] closes #25781
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -9416,6 +9417,10 @@ def corr(
             Minimum number of observations required per pair of columns
             to have a valid result. Currently only available for Pearson
             and Spearman correlation.
+        calculate_diagonal : bool, optional
+            Whether to calculate pairwise correlation using supplied callable.
+            Ignored when method argument is not callable. If False, pairwise
```
Comment (jreback) [inline pandas/core/frame.py:9422]: add a versionadded 1.4 tag

File: pandas/core/frame.py [hunk 2]
```diff
@@ -9416,6 +9417,10 @@ def corr(
             Minimum number of observations required per pair of columns
             to have a valid result. Currently only available for Pearson
             and Spearman correlation.
+        calculate_diagonal : bool, optional
```
Comment (jreback) [inline pandas/core/frame.py:9420]: this is descriptive, but is there precedent for this type of naming, e.g numpy?

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — COMMENTED** [review]
can you add a whatsnew note in other enhancements for 1.4

File: pandas/core/frame.py (hunk 2 above)
Comment (peterpanmj) [reply pandas/core/frame.py:9420]: No. I came up with this myself. I'm not familiar with Numpy naming convention. Any suggestion ?

**Review by jreback — APPROVED** [review]

**Comment by Dr-Irv** [comment]
What `corr()` provides is computation of the correlation matrix of a `DataFrame`, where each pair of columns (`Series`) of the `DataFrame` is then used to compute a correlation value.  For our defined correlation functions (`‘pearson’, ‘kendall’, ‘spearman’`), the correlation of two identical columns is defined as 1.  We also allow user-specified correlation functions.  In any case, if you have a `DataFrame` with `m` rows and `n` columns, the result is a `DataFrame` with `n` rows and `n` columns.   In our docs, we say that for a user-supplied correlation function, the diagonal entries will be 1.  

In the original issue, the user wanted to override the default value of 1. It seems to me there are a few ways of looking at this.  

1. Users want to use a "correlation" function that returns diagonal values not equal to 1.  One option is that the diagonal values should all be the same, in which case a keyword argument of `diagonal: float = 1.0` would make sense, and we stuff the same value on the diagonal.
2. Users want to use a "correlation" function that returns diagonal values not equal to 1.  A second option is that the diagonal values should be different, in which case a keyword argument of `compute_diagonal: bool= False` would make sense, and we let the supplied correlation function compute the diagonal (as in this PR)
3. One could argue that the only proper definition of a correlation function is that the diagonal values are 1.0.  Two identical `Series` are 100% correlated.  So our default value of 1.0 for a user-supplied function makes sense.  In that case, we let users override the values via something like:
```python
    result = df.corr(lambda x,y: myfunc(x,y))
    for i in range(len(df.columns)):
        result.iloc[i,i] = myfunc(df.iloc[:,i], df.iloc[:,i])
```
The latter seems a bit awkward, so I would disagree with @jbrockmendel where he says " it is just as easy to patch the result directly"

I think it is reasonable to ask @fabianrost84 (OP on the original issue) and @peterpanmj whether:
1. Do they want to supply one value for all the diagonals, or have different values for each pair?
2. Why is the workaround listed above using a loop not sufficient?

**Pull request closed without merging**